app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/OffboardRouterActivity.kt

```diff
@@ -0,0 +1,257 @@
+import com.mapbox.services.android.navigation.v5.navigation.metrics.MetricsObserver
+import com.mapbox.turf.TurfConstants
+import com.mapbox.turf.TurfMeasurement
+import kotlinx.android.synthetic.main.activity_mock_navigation.*
+import timber.log.Timber
+
+class OffboardRouterActivity : AppCompatActivity(),
+    OnMapReadyCallback,
+    MapboxMap.OnMapClickListener,
+    MetricsObserver,
+    DirectionsSession.RouteObserver {
+
+    private var mapboxMap: MapboxMap? = null
+
+    // Navigation related variables
+    private lateinit var locationEngine: LocationEngine
+    private lateinit var navigation: MapboxNavigation
+    private lateinit var navigationMapRoute: NavigationMapRoute
+    private var directionsSession: DirectionsSession? = null
+    private var route: DirectionsRoute? = null
+    private var destination: Point? = null
+    private var waypoint: Point? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_mock_navigation)
+
+        mapView.onCreate(savedInstanceState)
+        mapView.getMapAsync(this)
+
+        val customNotification = CustomNavigationNotification(applicationContext)
+        val options = MapboxNavigationOptions.Builder()
+            .navigationNotification(customNotification)
+            .defaultMilestonesEnabled(false)
+            .build()
+
+        navigation = MapboxNavigation(
+            this,
+            Utils.getMapboxAccessToken(this),
+            options
+        )
+
+        newLocationFab.setOnClickListener { onNewLocationClick() }
+    }
+
+    private fun newOrigin() {
+        mapboxMap?.let { map ->
+            clearMap(map)
+            val latLng = Utils.getRandomLatLng(doubleArrayOf(-77.1825, 38.7825, -76.9790, 39.0157))
+            (locationEngine as ReplayRouteLocationEngine).assignLastLocation(
+                Point.fromLngLat(latLng.longitude, latLng.latitude)
+            )
+            map.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 12.0))
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    override fun onMapReady(mapboxMap: MapboxMap) {
+        this.mapboxMap = mapboxMap
+        this.mapboxMap?.addOnMapClickListener(this)
+        mapboxMap.setStyle(Style.MAPBOX_STREETS) { style ->
+            val locationComponent = mapboxMap.locationComponent
+            val options = LocationComponentActivationOptions.builder(this, style).build()
+            locationComponent.activateLocationComponent(options)
+            locationComponent.renderMode = RenderMode.GPS
+            locationComponent.isLocationComponentEnabled = false
+            navigationMapRoute = NavigationMapRoute(navigation, mapView, mapboxMap)
+            Snackbar.make(findViewById(R.id.container), "Tap map to place waypoint", Snackbar.LENGTH_LONG).show()
+            locationEngine = ReplayRouteLocationEngine()
```

Comment:
Is this needed for this example? We're generating random points. Why do we need a `LocationEngine`?